Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm

```diff
@@ -401,34 +380,36 @@ static bool sampleBufferRenderersSupportKeySession()
 void SourceBufferPrivateAVFObjC::setTrackChangeCallbacks(const InitializationSegment& segment, bool initialized)
 {
     for (auto& videoTrackInfo : segment.videoTracks) {
-        videoTrackInfo.track->setSelectedChangedCallback([weakThis = WeakPtr { *this }, this, initialized] (VideoTrackPrivate& track, bool selected) {
-            if (!weakThis)
+        videoTrackInfo.track->setSelectedChangedCallback([weakThis = ThreadSafeWeakPtr { *this }, this, initialized] (VideoTrackPrivate& track, bool selected) mutable {
+            RefPtr protectedThis = weakThis.get();
+            if (!protectedThis)
                 return;
 
             if (initialized) {
                 trackDidChangeSelected(track, selected);
                 return;
             }
-            m_pendingTrackChangeTasks.append([weakThis, trackRef = Ref { track }, selected] {
-                if (weakThis)
-                    weakThis->trackDidChangeSelected(trackRef, selected);
+            m_pendingTrackChangeTasks.append([weakThis = WTFMove(weakThis), trackRef = Ref { track }, selected] {
+                if (RefPtr protectedThis = weakThis.get())
+                    protectedThis->trackDidChangeSelected(trackRef, selected);
             });
         });
     }
 
     for (auto& audioTrackInfo : segment.audioTracks) {
-        audioTrackInfo.track->setEnabledChangedCallback([weakThis = WeakPtr { *this }, this, initialized] (AudioTrackPrivate& track, bool enabled) {
-            if (!weakThis)
+        audioTrackInfo.track->setEnabledChangedCallback([weakThis = ThreadSafeWeakPtr { *this }, this, initialized] (AudioTrackPrivate& track, bool enabled) mutable {
+            RefPtr protectedThis = weakThis.get();
+            if (!protectedThis)
                 return;
 
             if (initialized) {
                 trackDidChangeEnabled(track, enabled);
                 return;
             }
 
-            m_pendingTrackChangeTasks.append([weakThis, trackRef = Ref { track }, enabled] {
-                if (weakThis)
-                    weakThis->trackDidChangeEnabled(trackRef, enabled);
+            m_pendingTrackChangeTasks.append([weakThis = WTFMove(weakThis), trackRef = Ref { track }, enabled] {
```

Comment:
Can setEnabledChangedCallback lambda be called several times?

Reply:
good point, particularly seeing that an incorrect move is what caused failure in the other MediaServerPrivate refcounted patch